### PR TITLE
Initial conversion for IPFS datastore interface

### PR DESCRIPTION
**Pull request by b5**
Story time!
*tl;dr; this sets us up to do all sorts of cool shit with IPFS down the road, and makes testing better now*

This PR originated with [failing tests on the identity server](https://circleci.com/gh/archivers-space/identity). Previously this code was all in one package, but needed to be broken up in order to support remote procedure calls between servers (and for better encapsulation). Before this separation the package used an actual Postgres database connection to run it's tests. I'm a fan of testing against a real database because it's closer to what happens in production. The issue is, go runs different package test suites in parallel, so when the tests are broken up into separate packages, the tests start manipulating entries in the same database at the same time, causing unpredictable results. There are a few ways to fix this:

1. Spin up a database for each test suite
2. Mock the database
3. Abstract the storage layer, moving the Postgres tests up into the same package

First one just feels gross and slow to me. The second is a viable alternative, but the overhead of maintaining mocked database tests is rather high. I've done it in the past using [data-dog's sql-mock](https://github.com/DATA-DOG/go-sqlmock) and found it irritating, and didn't weed out nasty production bugs specific to, say Postgres 9.6.

So I got to thinking about the third option, and I was all "well at some point we might want a caching layer" (something like redis) because already the urls table is around 7M entries and takes forever to respond, so it would make sense to start thinking in the abstract about writing to an abstract "store", instead of "Postgres", and then I was all WHAT IF THE STORE WAS A DISTRIBUTED INTERNETS!?

So, to be clear, in a lot of places writing to IPFS doesn't make any sense, but in some it really, really does. For example, our web crawler sentry currently crawls & writes to a Postgres database, which we're thinking will eventually get packaged up & written to IPFS, the next feature for sentry is crawling JSON-LD files like data.gov, this PR sets up a world where we're _crawling directly to the distributed web_, creating a distributed, continuously updated data.gov.

So in order to do that we would need a bridge between the [ipfs-datastore interface](https://github.com/ipfs/go-datastore) and SQL, which I've written a cursory version of here: [sql_datastore](https://github.com/archivers-space/go-sql-datastore). That one even has documentation. This means we'll have toll-free writing to IPFS for all of our stuff, and our tests can use the datastore.MapDatastore() for unit tests, and move the Postgres tests up to top level packages. I've intentionally made the the package very specific because I don't want to spend too much time on this right now, as the pifs-datastore interface itself may change in the future. This does, however, make me more suited to provide input to the Protocol Labs team on the interface itself, using real-world examples.  

This also means that we can use more off-the-shelf code that implements the ipfs-datastore interface, and contribute to those projects in the process.

So, why is this a terrible idea? Well, it complicates the code, which makes it harder for others to understand, but it is much easier to write tests for that code, because you no longer need a database to do so. So with this I'll do what I can to up my documentation of what's going on for others to be able to grok.

cc @flyingzumwalt

### CODE REVIEW

**Comment by b5** [comment]
Totally, only thing I'd add is this `sql_datastore` package is ready for use now. It may need improvement / enhancement down the road, but this PR passes the required tests to be put into use. I'm working on branches for the other services that adjust to the new datastore interface, converting a service is taking around 15 minutes, with another few hours of work reserved for the identity server (which needs more specific attention)

**Comment by titaniumbones** [comment]
thx.  updated my, er, "review".  not merging b/c I'd like to see if anyone with actual code knowledge has any concerns; but also feel comfortable leaving that decision up to @b5.

**Comment by dcwalk** [comment]
👍

I've read through the description and made my way (quickly) through the code--
in no way am I the person who should be doing code review so I will not merge in, comfortable with @b5 making that dec'n 

I mostly have a couple process/sideband q's (test coverage! local dev env! live date! etc...) that I'll take to `#dev-archivers-2`

**Comment by b5** [comment]
Thanks for the review team! Merging.
